### PR TITLE
Refactoring Scrolling Behaviour (WIP) 

### PR DESCRIPTION
**Pull request by zhouhua**

### CODE REVIEW

**Comment by itsonlyjames** [comment]
@zhouhua I've simplified the approach for `scrollTo` as well as adding a "Scroll Behaviour" toggle for "auto, smooth, and instant". It's in a sort of working state, it consistently scrolls to the correct spot and because it's not directly at top 0 (probably because of the inline title) it is shown underneath the sticky header. Need to test without inline title. 

Also we need to either add a prettier config or agree upon a code style, my editor defaults to double quotes and is throwing errors all over the place. Perhaps we can setup a prettier config file?

Keen to get your eyes over a few things and see if you can improve the consistency of scrolling to the top most header if at the bottom of the page with smooth scrolling enabled 👍

**Comment by itsonlyjames** [comment]
I also noticed the issue with the sticky header height not updating was because we called and set it **prior** to it changing, meaning that the reference was old and outdated. If we want to get the correct height, it's probably best to put it in a timeout after we run the scroll function.

**Comment by zhouhua** [comment]
@itsonlyjames Since CodeMirror-related APIs are not available in preview mode, I’ve decided to refactor the code over the next couple of days and perform some tedious calculations.

**Comment by itsonlyjames** [comment]
@zhouhua 🤦‍♂️ I did not even consider this when moving into the CodeMirror functionality... that is my bad. 

I'm actually going to investigate using Svelte for handling the rendering elements, it should assist in decoupling the functionality and UI elements. As well as it will allow for better styling and animation/fade in/out transitions. I will investigate this while you refactor the calculations.

**Comment by zhouhua** [comment]
Check out my latest commit; I've completed the calculation for title scroll top.

Introducing a UI framework is a good idea. Initially, I thought title rendering was relatively simple, so I directly manipulated the DOM. This approach negatively impacts both performance optimization and code readability.

**Comment by itsonlyjames** [comment]
Checking it out now. 

Have a look at my new branch, I've implemented Svelte and a store so we can simply push `headings.set([])` to render them in there. Will be great. https://github.com/zhouhua/obsidian-sticky-headings/commit/a3f90bde8a0019d2363f698886a608d64395fffc

File: src/getShownHeadings/source.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+export function getShownHeadingsSource(
+  headings: HeadingCache[],
+  view: MarkdownView,
+  stickyHeaderHeight: number,
+): { heading: HeadingCache; offset: number; }[] {
+  const scrollTop = view.editor.getScrollInfo().top;
+  const result: { heading: HeadingCache; offset: number; }[] = [];
+  for (const heading of headings) {
+    const { position } = heading;
+    const { top = 0 } = view.editMode.editor?.coordsAtPos({ line: position.start.line, ch: 0 }, true) || {};
```
Comment (itsonlyjames) [inline src/getShownHeadings/source.ts:12]: I think using `view.editMode.cm.lineBlockAt(position.start.offset)` will provide a more bulletproof solution as that accounts for elements not currently rendered outside of the viewport, but it does change the way the calculations are done. Depending upon testing it might be okay.

Comment (itsonlyjames) [reply src/getShownHeadings/source.ts:12]: Comes from here: https://discuss.codemirror.net/t/smooth-scroll-to-selection-position/4875/4

File: src/getShownHeadings/index.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+function getShownHeadings(
+  headings: HeadingCache[],
+  view: MarkdownView,
+  stickyHeaderHeight: number,
+  settings: ISetting,
+): { heading: HeadingCache; offset: number; }[] {
+  const { type } = view.currentMode;
+  const getValidHeadings = type === 'preview' ? getShownHeadingsPreview : getShownHeadingsSource;
+  const validHeadings = getValidHeadings(headings, view, stickyHeaderHeight);
```
Comment (itsonlyjames) [inline src/getShownHeadings/index.ts:49]: This works correctly after the first heading, for some reason I'm getting all returned with offset 0 on first scroll

Aug-17-2024 02-31-20

File: src/Plugin.ts [hunk 3]
```diff
@@ -221,33 +212,36 @@ export default class StickyHaeddingsPlugin extends Plugin {
...
-          // wait for headings tree rendered
-          // @ts-expect-error typing error
-          view.currentMode.applyScroll(heading.position.start.line, { highlight: true });
-        }, 20);
+        // check if live preview and use different scroller element
+        (view.currentMode.type === 'preview'
+          ? view.previewMode.renderer.previewEl
+          : view.editor.cm.scrollDOM)
+          .scrollTo({
+            top: offset - stickyContainerHeight,
```
Comment (itsonlyjames) [inline src/Plugin.ts:223]: For the most part, the calculations and everything are working and they're scrolling correctly. However occasionally it'll just shoot off randomly to near the top. See examples. 

Aug-17-2024 02-37-54
Aug-17-2024 02-38-47

Comment (itsonlyjames) [reply src/Plugin.ts:223]: I've also provided the culprit issue file in hopes you might be able to replicate: https://pastebin.com/AinG2RL5

**Review by itsonlyjames — CHANGES_REQUESTED** [review]
We're heading in the right direction 🙏
I'll update master prettier file to match yours, and I can look at investigating pushing the headers into my svelte component.

File: src/getShownHeadings/source.ts (hunk 1 above)
Comment (zhouhua) [reply src/getShownHeadings/source.ts]: I'm not entirely sure about the exact differences between the two methods. The reason I chose to use `editor.coordsAtPos` is because this method calculates based on the entire editor or window boundaries, including the space for inline title and the properties editor. On the other hand, `cm.lineBlockAt` calculates based on `.cm-contentContainer` boundaries, which do not include the space for inline title and the properties editor.

However, seeing that you have tested and found many cases where the title offset returns as 0, you might be right. `editor.coordsAtPos` might not be reliable in our scenario. I have already switched to using cm.lineBlockAt.

File: src/getShownHeadings/index.ts (hunk 2 above)
Comment (zhouhua) [reply src/getShownHeadings/index.ts:49]: I haven't encountered this issue before. I've modified the title position calculation as per your suggestion. Let's observe if this resolves the problem.

File: src/Plugin.ts (hunk 3 above)
Comment (zhouhua) [reply src/Plugin.ts:223]: I downloaded the file and couldn't reproduce your issue. From your first screenshot, it looks like the cursor's position was targeted. Did you press any keys, or is there a plugin modifying the file's content?

**Comment by zhouhua** [comment]
I have a favorite Obsidian theme called "Border." It's a bit unique in that the editor doesn't have a background color set, so our plugin appears like this:

image

It seems a bit challenging to completely match the title background color with the editor background color.

So I tried changing the style a bit. Do you think this style looks good?

image

File: src/Plugin.ts (hunk 3 above)
Comment (itsonlyjames) [reply src/Plugin.ts:223]: I run this all in a test vault that has minimal plugins so as to avoid any conflicts, but I'll chalk it up to an edge-case. I wonder if screen resolution becomes an issue? Maybe it's different on my macbook screen—high-res—compared to my monitor which is 1440p instead. I'll have to test.

**Comment by itsonlyjames** [comment]
I'm also a big fan of the Border theme. 

I think your implementation looks great! 

If you're happy to, we can get this PR ready to merge and then use my https://github.com/zhouhua/obsidian-sticky-headings/tree/feature/SvelteUI branch to do all the styling. I've completely refactored the entire plugin, however it will make it easier to work with moving forward.

File: src/getShownHeadings/index.ts (hunk 2 above)
Comment (itsonlyjames) [reply src/getShownHeadings/index.ts:49]: Issue resolved 🙏

**Review by itsonlyjames — APPROVED** [review]
